### PR TITLE
add support for existing tests after hf1270

### PR DESCRIPTION
**Pull request by oxarbitrage**
By comments in https://github.com/bitshares/bitshares-core/pull/1324#issuecomment-439715251 the flag for existing tests was added to database fixture.

This pull is not to merge(at least not as it is) but it is for discussion of some issues found in the existing tests after the hardfork.

Trying to give support for the following existing test cases:
- hardfork_core_338_test
- hardfork_core_453_test
- hardfork_core_625_big_limit_order_test
- target_cr_test_limit_call
- target_cr_test_call_limit

The first 3 tests pass after hardfork, only mod needed is on `hardfork_core_338_test` that use call price to do some checks. This is not allowed after HF1270, order matching is as expected.

In the last 2 there are some different at matching orders, specifically i modified a bit `target_cr_test_limit_call` to show this differences:

This is how before HF match the orders before and after the big sell order:

```
NAME               FOR SALE         FOR WHAT PRICE (S/W) 1/PRICE (W/S)
======================================================================
seller              7 USDBIT         78 BTS      0.08974   11.14286 
buyer3            111 BTS           10 USDBIT   11.10000    0.09009 
buyer2          33000 BTS         3000 USDBIT   11.00000    0.09091 
buyer              80 BTS           10 USDBIT    8.00000    0.12500 
NAME               FOR SALE         FOR WHAT PRICE (S/W) 1/PRICE (W/S)
======================================================================
seller              7 USDBIT         78 BTS      0.08974   11.14286 
buyer2          12056 BTS         1096 USDBIT   11.00000    0.09091 
buyer              80 BTS           10 USDBIT    8.00000    0.12500 
```

And this is what after HF will do:

```
NAME               FOR SALE         FOR WHAT PRICE (S/W) 1/PRICE (W/S)
======================================================================
seller              7 USDBIT         78 BTS      0.08974   11.14286 
buyer3            111 BTS           10 USDBIT   11.10000    0.09009 
buyer2          33000 BTS         3000 USDBIT   11.00000    0.09091 
buyer              80 BTS           10 USDBIT    8.00000    0.12500 
NAME               FOR SALE         FOR WHAT PRICE (S/W) 1/PRICE (W/S)
======================================================================
seller              7 USDBIT         78 BTS      0.08974   11.14286 
buyer2          24310 BTS         2210 USDBIT   11.00000    0.09091 
buyer              80 BTS           10 USDBIT    8.00000    0.12500 
```
@abitmore i will like to have your word about this when you can.

### CODE REVIEW

**Comment by abitmore** [comment]
@oxarbitrage could you print data of the call orders? There should be a difference.

**Comment by oxarbitrage** [comment]
before HF call orders:

```
NAME             TYPE             DEBT           COLLAT  CALL PRICE(D/C) ~CALL PRICE(C/D)        SWAN(D/C)        SWAN(C/D)
======================================================================
borrower        1000 USDBIT        15000 BTS          8.57143          0.11667          0.06667         15.00000 
borrower2       1000 USDBIT        15500 BTS          8.85714          0.11290          0.06452         15.50000 
borrower3       1000 USDBIT        25000 BTS         14.28571          0.07000          0.04000         25.00000 
NAME             TYPE             DEBT           COLLAT  CALL PRICE(D/C) ~CALL PRICE(C/D)        SWAN(D/C)        SWAN(C/D)
======================================================================
borrower         615 USDBIT        10765 BTS         10.00232          0.09998          0.05713         17.50407 
borrower2        499 USDBIT         9989 BTS         11.43888          0.08742          0.04995         20.01804 
borrower3       1000 USDBIT        25000 BTS         14.28571          0.07000          0.04000         25.00000 
```

After:

```
NAME             TYPE             DEBT           COLLAT  CALL PRICE(D/C) ~CALL PRICE(C/D)        SWAN(D/C)        SWAN(C/D)
======================================================================
borrower        1000 USDBIT        15000 BTS          1.00000          1.00000          0.06667         15.00000 
borrower2       1000 USDBIT        15500 BTS          1.00000          1.00000          0.06452         15.50000 
borrower3       1000 USDBIT        25000 BTS          1.00000          1.00000          0.04000         25.00000 
NAME             TYPE             DEBT           COLLAT  CALL PRICE(D/C) ~CALL PRICE(C/D)        SWAN(D/C)        SWAN(C/D)
======================================================================
borrower3       1000 USDBIT        25000 BTS          1.00000          1.00000          0.04000         25.00000 
```

**Comment by abitmore** [comment]
The issue is fixed in `1270-fix-call-price-cache` branch. Thanks.

**Comment by oxarbitrage** [comment]
thank you @abitmore . Everything passing now, i removed the prints at https://github.com/bitshares/bitshares-core/pull/1493/commits/3aec53d9266c33e265e8aabad7348d59d70e0308 . Can be merged as it is.

Do you think we should do more HF1270 checks inside rounding_tests, bitassets_tests, swan tests?

**Comment by abitmore** [comment]
Yes, we need swan_tests (including revivals). I guess we need bitasset_tests as well, but perhaps don't need to update rounding tests. Thanks.

**Comment by abitmore** [comment]
@oxarbitrage that's an old test case which tests behaviors before any recent hard fork. I guess the behavior has changed since 2.0.180612, perhaps change the test case to that date and see what will happen?

File: tests/tests/bitasset_tests.cpp [hunk 1]
```diff
@@ -940,6 +944,10 @@ BOOST_AUTO_TEST_CASE( hf_935_test )
          generate_blocks( HARDFORK_CORE_935_TIME - mi, true, skip );
          generate_blocks( db.get_dynamic_global_properties().next_maintenance_time, true, skip );
       }
+      else if( i == 6 ) // go beyond hard fork 1270
+      {
+         generate_blocks( HARDFORK_CORE_1270_TIME, true, skip );
```
Comment (abitmore) [inline tests/tests/bitasset_tests.cpp]: To be more accurate, here should be
```
         generate_blocks(HARDFORK_CORE_1270_TIME - mi, true, skip);
         generate_blocks(db.get_dynamic_global_properties().next_maintenance_time, true, skip);
```

File: tests/tests/bitasset_tests.cpp [hunk 2]
```diff
@@ -1103,22 +1111,30 @@ BOOST_AUTO_TEST_CASE( hf_935_test )
...
-         // the limit order should have been filled
-         // TODO FIXME this test case is failing for MCR test,
-         //            because call_order's call_price didn't get updated after MCR changed
-         // BOOST_CHECK( !db.find<limit_order_object>( sell_id ) );
-         if( i % 2 == 1 ) // MSSR test
-            BOOST_CHECK( !db.find<limit_order_object>( sell_id ) );
+         if( i % 2 != 0 ) { //  MSSR test, MSSR should be 125%, order filled
+            BOOST_CHECK_EQUAL(usd_id(db).bitasset_data(db).current_feed.maximum_short_squeeze_ratio, 1250);
+            BOOST_CHECK(!db.find<limit_order_object>(sell_id));
+         }
```
Comment (abitmore) [inline tests/tests/bitasset_tests.cpp]: Why removed the MCR test? To be consistent, I'd prefer putting MCR test in `if` and MSSR test in `else`.
```
if (mcr_test) // MCR test, MCR should be 350%, but due to issue #1270, call_order's call_price won't change, so the order is not filled
{
   BOOST_CHECK_EQUAL( usd_id(db).bitasset_data(db).current_feed.maintenance_collateral_ratio, 3500 );
   BOOST_CHECK( db.find<limit_order_object>( sell_id ) );
   // can also check whether call_price has changed
}
else // mssr test
{
            BOOST_CHECK_EQUAL(usd_id(db).bitasset_data(db).current_feed.maximum_short_squeeze_ratio, 1250);
            BOOST_CHECK(!db.find<limit_order_object>(sell_id));
}
```

File: tests/tests/bitasset_tests.cpp [hunk 3]
```diff
@@ -1103,22 +1111,30 @@ BOOST_AUTO_TEST_CASE( hf_935_test )
...
+         // go beyond hard fork 1270
+         blocks += generate_blocks(HARDFORK_CORE_1270_TIME, true, skip);
```
Comment (abitmore) [inline tests/tests/bitasset_tests.cpp]: To be more accurate, here should be
```
         blocks += generate_blocks(HARDFORK_CORE_1270_TIME - mi, true, skip);
         blocks += generate_blocks(db.get_dynamic_global_properties().next_maintenance_time, true, skip);
```

File: tests/tests/market_tests.cpp [hunk 4]
```diff
@@ -231,9 +231,11 @@ BOOST_AUTO_TEST_CASE(issue_338_etc)
...
 BOOST_AUTO_TEST_CASE(hardfork_core_338_test)
 { try {
-   auto mi = db.get_global_properties().parameters.maintenance_interval;
-   generate_blocks(HARDFORK_CORE_343_TIME - mi); // assume all hard forks occur at same time
-   generate_blocks(db.get_dynamic_global_properties().next_maintenance_time);
+
+   if(hf1270)
+      generate_blocks(HARDFORK_CORE_1270_TIME);
+   else
+      generate_blocks(HARDFORK_CORE_343_TIME);
```
Comment (abitmore) [inline tests/tests/market_tests.cpp]: To be more accurate, here we should generate blocks until the maintenance time but not the hf time.

File: tests/tests/market_tests.cpp [hunk 5]
```diff
@@ -406,9 +410,11 @@ BOOST_AUTO_TEST_CASE(hardfork_core_338_test)
...
 BOOST_AUTO_TEST_CASE(hardfork_core_453_test)
 { try {
-   auto mi = db.get_global_properties().parameters.maintenance_interval;
-   generate_blocks(HARDFORK_CORE_453_TIME - mi); // assume all hard forks occur at same time
-   generate_blocks(db.get_dynamic_global_properties().next_maintenance_time);
+
+   if(hf1270)
+      generate_blocks(HARDFORK_CORE_1270_TIME);
+   else
+      generate_blocks(HARDFORK_CORE_343_TIME);
```
Comment (abitmore) [inline tests/tests/market_tests.cpp]: To be more accurate, here we should generate blocks until the maintenance time but not the hf time.

File: tests/tests/market_tests.cpp [hunk 6]
```diff
@@ -478,17 +484,18 @@ BOOST_AUTO_TEST_CASE(hardfork_core_453_test)
...
 BOOST_AUTO_TEST_CASE(hardfork_core_625_big_limit_order_test)
 { try {
-   auto mi = db.get_global_properties().parameters.maintenance_interval;
-   generate_blocks(HARDFORK_CORE_625_TIME - mi); // assume all hard forks occur at same time
-   generate_blocks(db.get_dynamic_global_properties().next_maintenance_time);
+
+   if(hf1270)
+      generate_blocks(HARDFORK_CORE_1270_TIME);
+   else
+      generate_blocks(HARDFORK_CORE_625_TIME);
```
Comment (abitmore) [inline tests/tests/market_tests.cpp]: To be more accurate, here we should generate blocks until the maintenance time but not the hf time.

File: tests/tests/market_tests.cpp [hunk 7]
```diff
@@ -1195,9 +1202,11 @@ BOOST_AUTO_TEST_CASE(hard_fork_343_cross_test)
...
 BOOST_AUTO_TEST_CASE(target_cr_test_limit_call)
 { try {
-   auto mi = db.get_global_properties().parameters.maintenance_interval;
-   generate_blocks(HARDFORK_CORE_834_TIME - mi);
-   generate_blocks(db.get_dynamic_global_properties().next_maintenance_time);
+
+   if(hf1270)
+      generate_blocks(HARDFORK_CORE_1270_TIME);
+   else
+      generate_blocks(HARDFORK_CORE_834_TIME);
```
Comment (abitmore) [inline tests/tests/market_tests.cpp]: To be more accurate, here we should generate blocks until the maintenance time but not the hf time.

File: tests/tests/market_tests.cpp [hunk 8]
```diff
@@ -1373,9 +1382,11 @@ BOOST_AUTO_TEST_CASE(target_cr_test_limit_call)
...
 BOOST_AUTO_TEST_CASE(target_cr_test_call_limit)
 { try {
-   auto mi = db.get_global_properties().parameters.maintenance_interval;
-   generate_blocks(HARDFORK_CORE_834_TIME - mi);
-   generate_blocks(db.get_dynamic_global_properties().next_maintenance_time);
+
+   if(hf1270)
+      generate_blocks(HARDFORK_CORE_1270_TIME);
+   else
+      generate_blocks(HARDFORK_CORE_834_TIME);
```
Comment (abitmore) [inline tests/tests/market_tests.cpp]: To be more accurate, here we should generate blocks until the maintenance time but not the hf time.

File: tests/tests/swan_tests.cpp [hunk 9]
```diff
@@ -114,6 +114,10 @@ struct swan_fixture : database_fixture {
       generate_blocks( HARDFORK_CORE_216_TIME );
       generate_block();
     }
+    void wait_for_hf_core_1270() {
+       generate_blocks( HARDFORK_CORE_1270_TIME );
+       generate_block();
+    }
```
Comment (abitmore) [inline tests/tests/swan_tests.cpp]: To be more accurate, here we should generate blocks until the maintenance time but not the hf time.

**Comment by abitmore** [comment]
By the way, similar code added in previous PR need to be updated as well, E.G.
```
BOOST_AUTO_TEST_CASE(mcr_bug_increase_before1270)
{ try {

   generate_blocks(HARDFORK_CORE_453_TIME);

   generate_block();

```
and also mcr_bug_increase_after1270, mcr_bug_decrease_before1270 and mcr_bug_decrease_after1270.

To be more accurate, should generate blocks until the maintenance time but not the hf time.

Actually, theoretically some test cases will fail if

File: tests/tests/bitasset_tests.cpp (hunk 3 above)
Comment (oxarbitrage) [reply tests/tests/bitasset_tests.cpp]: thank you for the suggestion, i actually tried to make it like that everywhere but there are some problems not sure why.

in the case of this particular test(`hf_935_test`) when i make the generate block chains i get error:

```
83898ms th_a       bitasset_tests.cpp:1150       test_method          ] 10 assert_exception: Assert Exception
prev: popping block would leave head block null
    {}
    th_a  fork_database.cpp:41 pop_block

    {}
    th_a  db_block.cpp:477 pop_block
```

**Comment by oxarbitrage** [comment]
The `bitasset_tests/hf_935_test` is a bit weird to me, maybe i am not understanding it correctly but `HARDFORK_CORE_868_890_TIME`, `HARDFORK_CORE_935_TIME`, `HARDFORK_CORE_343_TIME` are all the same date. 
Independently from changes of this pull request, with all those hardforks at the same time the block that tests beyond hard fork 890 and before hard fork 935 was never happening.

In this commit: https://github.com/bitshares/bitshares-core/pull/1493/commits/8466e198badff212eb95ad7254157630459630eb i removed that test and use the spot to test hardfork 1270. Iam unsure if it is the right thing to do, can revert back.

File: tests/tests/bitasset_tests.cpp [hunk 10]
```diff
@@ -928,26 +928,22 @@ BOOST_AUTO_TEST_CASE( hf_935_test )
...
       {
-         generate_blocks( HARDFORK_CORE_935_TIME - mi, true, skip );
-         generate_blocks( db.get_dynamic_global_properties().next_maintenance_time, true, skip );
-      }
-      else if( i == 6 ) // go beyond hard fork 1270
-      {
-         generate_blocks( HARDFORK_CORE_1270_TIME, true, skip );
+         generate_blocks( HARDFORK_CORE_1270_TIME - mi, true, skip );
       }
+      generate_blocks( db.get_dynamic_global_properties().next_maintenance_time, true, skip );
```
Comment (abitmore) [inline tests/tests/bitasset_tests.cpp]: Why this change? Only need to generate blocks to next maintenance time if `i` is in `{2,4,6}`.

File: tests/tests/bitasset_tests.cpp [hunk 11]
```diff
@@ -928,26 +928,22 @@ BOOST_AUTO_TEST_CASE( hf_935_test )
...
 
-   for( int i = 0; i < 8; ++i )
+   for( int i = 0; i < 6; ++i )
    {
       idump( (i) );
       int blocks = 0;
       auto mi = db.get_global_properties().parameters.maintenance_interval;
 
-      if( i == 2 ) // go beyond hard fork 890
+      if( i == 2 ) // go beyond hard fork 890, 343, 935
```
Comment (abitmore) [inline tests/tests/bitasset_tests.cpp]: This change is not correct, because in testnet hf935 and hf890 are not at the same time.

File: tests/tests/bitasset_tests.cpp [hunk 12]
```diff
@@ -1066,45 +1062,12 @@ BOOST_AUTO_TEST_CASE( hf_935_test )
...
-            BOOST_CHECK_EQUAL( usd_id(db).bitasset_data(db).current_feed.maintenance_collateral_ratio, 3500 );
-         else // MSSR test, MSSR should be 125%
-            BOOST_CHECK_EQUAL( usd_id(db).bitasset_data(db).current_feed.maximum_short_squeeze_ratio, 1250 );
-
-         if( affected_by_hf_343 ) // if updated bitasset before hf 890, and hf 343 executed after hf 890
-            // the limit order should have been filled
-            BOOST_CHECK( !db.find<limit_order_object>( sell_id ) );
-         else // if not affected by hf 343
-            // the limit order should be still there, because `check_call_order` was incorrectly skipped
-            BOOST_CHECK( db.find<limit_order_object>( sell_id ) );
```
Comment (abitmore) [inline tests/tests/bitasset_tests.cpp]: I think this change is also not correct.

File: tests/tests/bitasset_tests.cpp (hunk 11 above)
Comment (oxarbitrage) [reply tests/tests/bitasset_tests.cpp]: they are actually at the same time that's why the changes:

https://github.com/bitshares/bitshares-core/blob/master/libraries/chain/hardfork.d/CORE_935.hf
https://github.com/bitshares/bitshares-core/blob/master/libraries/chain/hardfork.d/CORE_868_890.hf

also:

https://github.com/bitshares/bitshares-core/blob/master/libraries/chain/hardfork.d/CORE_343.hf

**Comment by abitmore** [comment]
The logic around `i` is as follows (code executes in order):
* the 1st pass, no block generation at the beginning, do something related to MCR, go beyond hf343, do something, go beyond hf 935, do something, go beyond hf1270, pop all blocks (roll back all changes)
* the 2nd pass, mostly same as 1st, the only difference is to test MSSR but not MCR, after this all changes are rolled back
* the 3rd pass, generate blocks to the time right after hf343, do something related to MCR, then go beyond hf935 and hf1270, pop blocks until the time right after hf343, so at this stage the chain is after hf343
* the 4th pass, we're now after hf343, do tests related to MSSR, go beyond hf935 and hf1270, roll back to hf343
* the 5th pass, generate blocks to hf935, test MCR, roll back to the time right after hf 935
* the 6th pass, test MSSR, roll back to hf935
* the 7th pass, generate blocks to hf1270, test MCR, roll back to the time right after hf1270
* the 8th pass, test MSSR, roll back to hf1270

File: tests/tests/bitasset_tests.cpp [hunk 13]
```diff
@@ -1050,79 +1054,64 @@ BOOST_AUTO_TEST_CASE( hf_935_test )
...
+         // go beyond hard fork 935
+         blocks += generate_blocks(HARDFORK_CORE_935_TIME - mi, true, skip);
+         blocks += generate_blocks(db.get_dynamic_global_properties().next_maintenance_time, true, skip);
       }
 
-      // after hard fork 890, if it's before hard fork 935
-      if( db.get_dynamic_global_properties().next_maintenance_time <= HARDFORK_CORE_935_TIME )
+      // after hard fork 935, the limit order is filled only for the MSSR test
+      if( db.get_dynamic_global_properties().next_maintenance_time > HARDFORK_CORE_935_TIME &&
+          db.get_dynamic_global_properties().next_maintenance_time <= HARDFORK_CORE_1270_TIME && i > 4)
```
Comment (abitmore) [inline tests/tests/bitasset_tests.cpp]: `i > 4` should not be here. We still need the checks when `i <= 4`.

File: tests/tests/bitasset_tests.cpp (hunk 11 above)
Comment (abitmore) [reply tests/tests/bitasset_tests.cpp]: Please also check `testnet` branch.

**Comment by oxarbitrage** [comment]
thank you, with the logic described i should be able to make it better.

File: tests/tests/bitasset_tests.cpp [hunk 14]
```diff
@@ -1103,22 +1118,39 @@ BOOST_AUTO_TEST_CASE( hf_935_test )
...
-            BOOST_CHECK_EQUAL( usd_id(db).bitasset_data(db).current_feed.maximum_short_squeeze_ratio, 1250 );
-         // the limit order should have been filled
-         // TODO FIXME this test case is failing for MCR test,
-         //            because call_order's call_price didn't get updated after MCR changed
-         // BOOST_CHECK( !db.find<limit_order_object>( sell_id ) );
-         if( i % 2 == 1 ) // MSSR test
-            BOOST_CHECK( !db.find<limit_order_object>( sell_id ) );
+         if( i % 2 == 0) { // MCR test, median MCR should be 350% and order will not be filled except when i = 0
+            BOOST_CHECK_EQUAL(usd_id(db).bitasset_data(db).current_feed.maintenance_collateral_ratio, 3500);
+            if(i / 2 == 0)  // order is filled with i = 0 at HARDFORK_CORE_343_TIME + mi
```
Comment (abitmore) [inline tests/tests/bitasset_tests.cpp]: Should replace this with `if( affected_by_hf_343 )`.

File: tests/tests/market_tests.cpp [hunk 15]
```diff
@@ -1283,7 +1308,7 @@ BOOST_AUTO_TEST_CASE(target_cr_test_limit_call)
    BOOST_CHECK_LT( call2_to_cover.value, call2_id(db).debt.value );
    // even though call2 has a higher CR, since call's TCR is less than call2's TCR, so we expect call will cover less when called
    BOOST_CHECK_LT( call_to_cover.value, call2_to_cover.value );
-
+
```
Comment (abitmore) [inline tests/tests/market_tests.cpp]: Please remove the trailing white spaces.
